### PR TITLE
[FLINK-28433][connector/jdbc]Add mariadb jdbc connection validation

### PR DESCRIPTION
**Pull request by bzhaoopenstack**
## What is the purpose of the change

1. Flink Connector support the jdbc connection with Mysql series DB, such as MySQL, MariaDB,
Percona and etc. But we see the doc reference[1] shows the Mariadb Connector/J 3.0 will
accepts jdbc:mariadb: as the protocal in connection strings by default. So we need to apply
this change in our mysql jdbc connection vailidation. Mariadb Connector/J 2.X still support
both "jdbc:mysql:" and "jdbc:mariadb:".
2. Introduce MariaDB dialect implemented like Mysql.
3. Add UTs for DataTypes and a testcontainer for MariaDB in
   JdbcExactlyOnceSinkE2eTest.

[1] https://mariadb.com/kb/en/about-mariadb-connector-j/#jdbcmysql-scheme-compatibility

## Brief change log

-  Making Mysql jdbc connector validate the MariaDB type connection protocol for fit users to use Mysql jdbc connector to access MariaDB instances.
- Introduce MariaDB jdbc dialect into Flink, which is very similar with Mysql.
- Introduce UTs and E2E test for them.

## Verifying this change

Exec `mvn clean package -Dfast -Pskip-webui-build -pl flink-connectors/flink-connector-jdbc`
Or `mvn clean package -Dfast -Pskip-webui-build -pl flink-connectors/flink-connector-jdbc -Dtest="org.apache.flink.connector.jdbc.xa.JdbcExactlyOnceSinkE2eTest"` for run all tests or a single test. 
Notes that all types didn't cover, and will be on going for support.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`:  (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive):  (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (no) 
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)
  - If yes, how is the feature documented? (docs)

### CODE REVIEW

**Comment by Peng-Lei** [comment]
+1， After the change. We can use mariadb driver to connect the mysql through by setting the jdbcOptions of `driver` parameter without change the url prefix.

**Review by MartijnVisser — CHANGES_REQUESTED** [review]
@bzhaoopenstack Thanks for the PR, but in order to add support for MariaDB I would also like to see a test and documentation in place.

**Comment by bzhaoopenstack** [comment]
> @bzhaoopenstack Thanks for the PR, but in order to add support for MariaDB I would also like to see a test and documentation in place.

Thanks very much for reply and suggest. I will add the test and documentation. Once I'm finish, I think I need your kind review and advice. Thank you. ;-)

**Comment by bzhaoopenstack** [comment]
> @bzhaoopenstack Thanks for the PR, but in order to add support for MariaDB I would also like to see a test and documentation in place.

Hi @MartijnVisser, I have a question here, you said you want a test for this PR, does it mean we should add a specific UT which only verify the jdbc conncetion strings?  Because I didn't find any existing UT in Flink.
And notes: We only use Mysql connector to connect MariaDB, as it is a Mysql-series database.

**Comment by MartijnVisser** [comment]
I would expect that we add MariaDB to the `JdbcExactlyOnceSinkE2eTest` with the correct test container and have a MariaDB dialect implemented (even if that would be just a copy of MySQL)

**Comment by bzhaoopenstack** [comment]
> I would expect that we add MariaDB to the `JdbcExactlyOnceSinkE2eTest` with the correct test container and have a MariaDB dialect implemented (even if that would be just a copy of MySQL)

Thanks for explain more, this is a good direction/suggestion for me.

**Comment by bzhaoopenstack** [comment]
Hi @MartijnVisser , sorry for late.
And this is my first PR in flink, so I wish you could leave some kind directions and suggestions. Thank you very much. ;-). And hope this doesn't disturb you so much.

Now, I will request an new review from your side, once we both think this PR is ready, I will remove the title [WIP]. ;-)

File: flink-connectors/flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/dialect/mysql/MySqlDialectFactory.java [hunk 1]
```diff
@@ -27,7 +27,7 @@
 public class MySqlDialectFactory implements JdbcDialectFactory {
     @Override
     public boolean acceptsURL(String url) {
-        return url.startsWith("jdbc:mysql:");
+        return (url.startsWith("jdbc:mysql:") || url.startsWith("jdbc:mariadb:"));
```
Comment (Peng-Lei) [inline flink-connectors/flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/dialect/mysql/MySqlDialectFactory.java:30]: I think no need change here after we add `MariaDBDialectFactory`

**Comment by MartijnVisser** [comment]
@afedulov Do you want to have a look at this one?

**Comment by bzhaoopenstack** [comment]
Hi all. @MartijnVisser @hadronzoo any updates on this one? Thanks

**Comment by Peng-Lei** [comment]
@bzhaoopenstack Are you ready？If yes, please change the title without "WIP".

**Comment by bzhaoopenstack** [comment]
> @bzhaoopenstack Are you ready？If yes, please change the title without "WIP".

Yeah, I'm ready for the kind suggest from @MartijnVisser and @hadronzoo . ;-)

**Comment by MartijnVisser** [comment]
@bzhaoopenstack I won't be able to review this at the moment, but keep in mind that the CI is still failing for this PR. It makes sense to make sure that the CI works before the review actually starts.

**Comment by bzhaoopenstack** [comment]
> @bzhaoopenstack I won't be able to review this at the moment, but keep in mind that the CI is still failing for this PR. It makes sense to make sure that the CI works before the review actually starts.

OK, thank you Martijn. I will resolve it and make it ready for review

**Comment by bzhaoopenstack** [comment]
Sorry for late update. ;-) . Let's wait the azure test result. I got pass on my local env.

**Comment by bzhaoopenstack** [comment]
@flinkbot run azure

**Comment by MartijnVisser** [comment]
We've now moved the code from the JDBC connector to its own dedicated repository; please re-route this PR to https://github.com/apache/flink-connector-jdbc

**Pull request closed without merging**